sklearn/preprocessing/data.py

```diff
@@ -3147,3 +3150,216 @@ def inverse_transform(self, X):
+    >>> enc.n_values_
+    array([2, 3, 4])
+    >>> enc.feature_indices_
+    array([0, 1, 3, 6])
+    >>> enc.transform([[0, 1, 2]])
+    array([[ 0.,  1.,  0.,  1.,  1.,  0.]])
+
+    See also
+    --------
+    sklearn.preprocessing.OneHotEncoder: encodes categorical integer features
+      using a one-hot aka one-of-K scheme.
+    """
+    def __init__(self, n_values="auto", ordinal_features="all",
+                 dtype=np.float64, sparse=False, handle_greater='warn'):
+        self.n_values = n_values
+        self.ordinal_features = ordinal_features
+        self.dtype = dtype
+        self.sparse = sparse
+        self.handle_greater = handle_greater
+
+    def fit(self, X, y=None):
+        """Fit UnaryEncoder to X.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_feature]
+            Input array of type int.
+            All feature values should be non-negative otherwise will raise a
+            ValueError.
+        """
+        _transform_selected(X, self._fit, self.ordinal_features, copy=True)
+        return self
+
+    def _fit(self, X):
+        """Assumes X contains only ordinal features."""
+        X = check_array(X, dtype=np.int)
+        if self.handle_greater not in ['warn', 'error', 'clip']:
+            raise ValueError("handle_greater should be either 'warn', 'error' "
+                             "or 'clip' got %s" % self.handle_greater)
+        if np.any(X < 0):
+            raise ValueError("X needs to contain only non-negative integers.")
+        n_samples, n_features = X.shape
+
+        if (isinstance(self.n_values, six.string_types) and
+                self.n_values == 'auto'):
+            n_values = np.max(X, axis=0) + 1
+        elif isinstance(self.n_values, numbers.Integral):
+            n_values = np.empty(n_features, dtype=np.int)
+            n_values.fill(self.n_values)
+        else:
+            try:
+                n_values = np.asarray(self.n_values, dtype=int)
+            except (ValueError, TypeError):
+                raise TypeError("Wrong type for parameter `n_values`. Expected"
+                                " 'auto', int or array of ints, got %r"
+                                % self.n_values)
+            if n_values.ndim < 1 or n_values.shape[0] != X.shape[1]:
+                raise ValueError("Shape mismatch: if n_values is an array,"
+                                 " it has to be of shape (n_features,).")
+
+        self.n_values_ = n_values
+        n_values = np.hstack([[0], n_values - 1])
+        indices = np.cumsum(n_values)
+        self.feature_indices_ = indices
+
+        mask = (X >= self.n_values_).ravel()
+        if np.any(mask):
+            if self.handle_greater == 'error':
+                raise ValueError("handle_greater='error' but found %d feature"
+                                 " values which exceeds n_values."
+                                 % np.count_nonzero(mask))
```

Comment:
In principle this check is not needed in case of `'auto'` (but not sure whether a small performance improvement for this case is worth moving it elsewhere)